File: example-oc/example-oc/DevUtil.m
//
//  DevUtil.m
//  example-oc
//
//  Created by 张磊 on 2017/10/26.
//  Copyright © 2017年 游艺春秋网络科技（北京）有限公司. All rights reserved.
//

#import "DevUtil.h"
#import <Foundation/Foundation.h>

// 实现对象
@implementation DevUtil;

// 全局变量
static int _acctGameUserId = 0;

/**
 获得游戏账号标识（临时变量）

 @return 值
 */
+ (int)acctGameUserId {
    return _acctGameUserId;
}

/**
 设置游戏账号标识（临时变量）

 @param num 值
 */
+ (void)setAcctGameUserId:(int)num {
    _acctGameUserId = num;
}

/**
 生成返回值结果集

 @param resultJSON 结果JSON字符串
 @return 参数辞典
 */
+ (NSDictionary<NSString *, id> *)genResultSet: (NSString *)resultJSON {
    NSDictionary *dic = [NSJSONSerialization JSONObjectWithData:[resultJSON dataUsingEncoding:NSUTF8StringEncoding]
                                                        options:0 error:nil];
    NSMutableDictionary *res = [NSMutableDictionary new];
    [res addEntriesFromDictionary:dic];
    NSString *str = [dic objectForKey:@"sdk_token"];
    if (str != nil) {
        NSArray *arr = [str componentsSeparatedByString:@"&"];
        for (NSString *pair in arr) {
            NSRange idx = [pair rangeOfString:@"="];
            if (idx.location == NSNotFound) {
                continue;
            }
            [res setValue:[pair substringFromIndex:idx.location+1]
                   forKey:[pair substringToIndex:idx.location]];
        }
    }
    return res;
}

/**
 模拟生成交易信息
 
 @param service 业务编号
 @param content 商品明细
 @param totalFee 付费总额
 @param acctGameUserId 账号标识
 @return 交易信息
 */
+ (NSString *)genTradeInfo: (int)service : (NSString *)content : (int)totalFee : (int)acctGameUserId {
    NSCharacterSet *allowedCharacters = [[NSCharacterSet characterSetWithCharactersInString:@"+=\"#%/<>?@\\^`{|} "] invertedSet];
    NSString *content_encoded = [content stringByAddingPercentEncodingWithAllowedCharacters:allowedCharacters];
    // 模拟服务端临时接口
    NSString *ifUrl = [NSString stringWithFormat:@"http://debug.sdk.m.iccgame.com/?module=GAME.Pays.TradeTest&do=Create&service=%d&content=%@&total_fee=%d&acct_game_user_id=%d",
                       service, content_encoded, totalFee, acctGameUserId];
    // 通过服务端获取数据
    NSData *jsonStr = [NSData dataWithContentsOfURL:[NSURL URLWithString:ifUrl] options:NSDataReadingUncached error:nil];
    NSDictionary<NSString *, id> *obj = [NSJSONSerialization JSONObjectWithData:jsonStr options:0 error:nil];
    NSArray *p0 = [obj objectForKey:@"data"];
    NSDictionary<NSString *, id> *p1 = p0[0];
    NSArray *p2 = [p1 objectForKey:@"args"];
    return p2[0];
}

@end

File: example/example/DevUtil.swift
//
//  DevUtil.swift
//  sdk
//
//  Created by 张磊 on 2017/10/19.
//  Copyright © 2017年 游艺春秋网络科技（北京）有限公司. All rights reserved.
//

import Foundation

/// 开发工具
public class DevUtil {

    public class ResultSet {
        public  init(resultJSON:String ) {
            
        }
    }
    
    /// 游戏账号标识（临时变量）
    public static var acctGameUserId = 0
    
    /// 生成结果集
    ///
    /// - Parameter resultJSON: 结果字符串
    /// - Returns: 词典对象
    public static func genResultSet(resultJSON:String) -> [String:Any] {
        var arr = try! JSONSerialization.jsonObject(with: resultJSON.data(using: .utf8)!) as! [String:Any]
        if arr.index(forKey: "sdk_token") != nil {
            let items = (arr["sdk_token"] as! String).components(separatedBy: "&")
            for str in items {
                let idx = str.range(of: "=")
                let key = str.substring(to: idx!.lowerBound)
                arr[key] = str.substring(from: idx!.upperBound)
            }
        }
        return arr
    }
   
    /// 模拟生成交易信息
    ///
    /// - Parameters:
    ///   - service: 业务编号
    ///   - content: 商品明细
    ///   - totalFee: 付费总额
    ///   - acctGameUserId: 账号标识
    /// - Returns: 交易信息
    public static func genTradeInfo (service:Int, content:String, totalFee:Int, acctGameUserId:Int) -> String {
        let allowedCharacters =  NSCharacterSet(charactersIn:"+=\"#%/<>?@\\^`{|} ").inverted
        let content_encoded = content.addingPercentEncoding(withAllowedCharacters: allowedCharacters)
        // 模拟服务端临时接口
        let ifUrl = String(format:"http://debug.sdk.m.iccgame.com/"
            + "?module=GAME.Pays.TradeTest"
            + "&do=Create"
            + "&service=%d"
            + "&content=%@"
            + "&total_fee=%d"
            + "&acct_game_user_id=%d",
                           service, content_encoded!, totalFee, acctGameUserId
        )
        // 通过服务端获取数据
        let jsonStr = try! Data(contentsOf: URL(string: ifUrl)!, options: .uncached)
        let obj = try? JSONSerialization.jsonObject(with: jsonStr) as! [String:Any]
        var p0 = obj?["data"] as! [Any]
        var p1 = p0[0] as! [String:Any]
        var p2 = p1["args"] as! [Any]
        return p2[0] as! String
    }

// end class
}
